### ISSUE TITLE
CircuitBreaker Trait

### ISSUE THREAD

**Issue by jakubjanecek** [monix/monix#1068]
Hello,
is there a reason why `CircuitBreaker` is not a `trait`? It would be very useful to have it as an interface with multiple possible implementation but mostly for the ability to wrap it and enrich it with extra features. What do you think? I could give it a try and implement it if this is something that could be done.

Thank you.

**Comment by Avasil** [comment]
Hello!

I'm not familiar with @alexandru thought process when he was implementing it. I suppose it's because it's just a data structure so a `trait` felt redundant

What other possible implementation do you have in mind? If there's a nice use case we could probably do this as long as it won't break bincompat

**Comment by jakubjanecek** [comment]
Well, the possibility of another implementation is just theoretical. My use case is to be able to wrap circuit breaker with some code, e.g. like this:

```scala
class LoggingCircuitBreaker[F[_]](cb: CircuitBreaker[F]) extends CircuitBreaker[F] {
  override def protect[A](task: F[A]): F[A] = {
    ...
    cb.protect(task)
    ...
}
```

Currently I am not able to do that.

Actually my use case is to be able to add logging and metrics to circuit breaker and this #1067 would be sufficient for me. But I think it would be much better design to have a `trait` - for wrapping, for testing, for anything else.

**Comment by jakubjanecek** [comment]
So I just encountered another use case. I'd like to enhance `CircuitBreaker` (wrap it again) with some logic in `onErrorRecoverWith` and without a `trait` I am not able to do that.

**Comment by Avasil** [comment]
Yeah I think wrapping would also allow to easily implement `onProtect` on user side.
I don't see anything wrong with trait / abstract class, e.g. `MVar` in Cats-Effect is like that.

I feel like wrapping is possible even now, just without common trait but I can see how it could be annoying if you want to use vanilla implementation in tests.

IMO go for it :)

**Comment by jakubjanecek** [comment]
I've given it a try in #1069. There is a problem with backwards compatibility so it's not really possible to introduce `CircuitBreaker` trait and have it clean. So I at least added `ICircuitBreaker` trait - this way it's backwards compatible but whoever wants can use `ICircuitBreaker` instead of `CircuitBreaker` directly.

**Comment by alexandru** [comment]
Hello,

Sorry for my latency, I've been busy at work and with life again 🙂 

@jakubjanecek to expand on [my comment](https://github.com/monix/monix/pull/1069#issuecomment-549743182) ...

> Could you give me an example why not? As described in #1068 I want to enrich the circuit breaker with logging and metrics and having a trait is very good for that. Also testing piece of code with circuit breaker would be much easier if you can mock it easily. Or is there another way I am missing?

A very practical reason for why not is because traits are terrible for preserving backwards compatibility.

When you make something a `trait`, going forward the only way to add methods to it is to describe a default implementation in terms of the available methods or `F[_]` restrictions. And if you don't want something that can be described in terms of the available methods or restrictions, you're out of luck.

We've seen this in Cats-Effect, where `Ref`, `Defer` and `MVar` are described as traits. Here is one such example of a PR that's been frozen:

https://github.com/typelevel/cats-effect/pull/367

Ironically this example is a straightforward problem that should have been solvable with a default implementation, but cannot be, because the restrictions for `F[_]` are not specified by the `trait`, but by the constructor — this was by design, however I now believe specifying an `F[_] : Monad` as a restriction would have been better.  This is the kind of compromise you end up doing when designing `traits` ...

*The more specific you are, the less generic that `trait` is, however the more friendly it is to future developments* ... pick your poison ;-)

Note in Cats-Effect exposing abstract stuff is acceptable, because Cats-Effect is meant to provide APIs that can be implemented by others. So going to such pains was maybe worth it, although the only known third-party implementations of those interfaces are in Monix (the already mentioned `MVar`), so even for Cats-Effect we have to ask ourselves if it was worth it.

I basically subscribe to the philosophy that all classes and methods should be final, unless there are really good reasons to make them virtual. There are 2 schools of thought here and I think the opinion of Anders Hejlsberg, the designer of C#, does more justice to this, see:

https://www.artima.com/intv/nonvirtual.html

> When we make something virtual in a platform, we're making an awful lot of promises about how it evolves in the future. For a non-virtual method, we promise that when you call this method, x and y will happen. When we publish a virtual method in an API, we not only promise that when you call this method, x and y will happen. We also promise that when you override this method, we will call it in this particular sequence with regard to these other ones and the state will be in this and that invariant.
>
> Every time you say virtual in an API, you are creating a call back hook. As an OS or API framework designer, you've got to be real careful about that. You don't want users overriding and hooking at any arbitrary point in an API, because you cannot necessarily make those promises. And people may not fully understand the promises they are making when they make something virtual.

And I also subscribe to the philosophy that breaking people's code by evolving APIs in backwards incompatible ways is bad and we should feel bad whenever we do it ... yes, I know we've done it in the past in Monix, but we should do that less in the future.

Rich Hickey has a wonderful presentation about it: https://www.youtube.com/watch?v=oyLBGkS5ICk

TL;DR — when you break APIs, you should change the namespace too (i.e. the package and the artifact's name). Of course, I'm not sure if we can do this in Scala, because we depend on others as well. But I'm seriously thinking of a future `monix.v4` namespace for when we'll break backwards compatibility again.

---

So that's basically the rub ... when you expose a `trait` instead of a plain class, you're making a promise that its API won't evolve. And I'm not sure if making such a promise is worth it.

**Comment by alexandru** [comment]
On your use-case ...

```
class LoggingCircuitBreaker[F[_]](cb: CircuitBreaker[F]) extends CircuitBreaker[F] {
  override def protect[A](task: F[A]): F[A] = {
    ...
    cb.protect(task)
    ...
}
```

As a personal opinion I don't think an `ICircuitBreaker` buys you much, because then you'd need to implement all of those methods by yourself and that's boilerplate, plus in actual code you only initialize the `CircuitBreaker` once and then use its `protect` everywhere else. This means that, if you want subtyping polymorphism, you could have your own interface, without all the junk ...

```scala
trait MyCircuitBreaker[F[_]] {
  def protect[A](task: F[A]): F[A]
}
```

**Comment by jakubjanecek** [comment]
@alexandru Thank you for your extensive explanation. I did not have to deal with backwards compatibility much (internal projects) in the past so I don't have that experience and I am used to have `trait`s basically everywhere. I wonder, don't you have the same problem with the current implementation though? If you decide to somehow extend the `CircuitBreaker` class with new public method you will have to also implement it to keep backwards compatibility, right? Won't you encounter the same set of problems described for traits?

The problem with my own `MyCircuitBreaker` interface is that I am actually writing a library and I don't want to provide my own circuit breaking interface, I want this library to be compatible with Monix `CircuitBreaker`. Anything can be wrapped in my own wrapper but that is acceptable only in some application not a library. But I do want to extend it with logging and metrics which are "missing" from the current implementation. To be really concrete I am trying to do this: https://github.com/avast/scala-server-toolkit/blob/master/monix-catnap/src/main/scala/com/avast/sst/monix/catnap/CircuitBreakerModule.scala#L53 but right now I am not able to measure how many requests went through the circuit breaker (I just have the number of rejected and the current state).

**Comment by jakubjanecek** [comment]
@alexandru I'll try to explain my exact use case again and propose possible solutions because I see this PR won't be merged in the current state.

I am writing a library which helps you initialize components from other well-known libraries (such as `CircuitBreaker` from Monix) and sometimes enriches them with extra features. The library does not want to provide its own abstractions (such as `MyCircuitBreaker`), really the only purpose is to simplify initialization and make it a bit consistent across different libraries. Whenever we have some HTTP client we want to protect it by `CircuitBreaker` but we also want to add important logging and metrics (such as when the circuit is closed, when it rejects something, ...). So what I would like to do is to initialize a `CircuitBreaker` and enrich it with the logging and metrics and return that enriched instance to the user. As I said I don't want to provide my own interface - that would be directly against my goals. The current implementation of `CircuitBreaker` does not alllow me to do some action every time `protect` is called (e.g. to log some message and increase counter of requeests). Of course there are ways to do it but what I am talking about is being able to do so seamlessly without the user knowing. The user just gets `CircuitBreaker` instance and it protects him, logs messages and has some metrics. I know it might seem like I'm mixing too many things together but I know from experience that this is very important for production monitoring and very valuable during debugging problems.

I see two possible solutions:
* There is a `trait` such as `ICircuitBreaker` which allows me to wrap `CircuitBreaker` and enrich it.
* There is method `onProtect` in `CircuitBreaker` (#1067).

Both of these are possible for my use case even though I still think some kind of circuit breaker interface would be valuable. I've tried to implement `onProtect` but I don't think it's possible to keep backwards compatibility too with this approach (more specifically it's ok in `CircuitBreaker` but there is problem with compatibility in the `Builders` class).

I think that maybe I've made the `ICircuitBreaker` interface too broad. I copied all the public methods of `CircuitBreaker` but what would you say if we stripped it down to the essence of circuit breaking which is:

```scala
trait ICircuitBreaker[F[_]] {
  def protect[A](task: F[A]): F[A]
}
```

It seems to me that we could have such essential interface and be sure that we won't need to extend it in the future which means it's quite safe to have. Someone might suggest to get rid of this simple interface and just use `F[A] => F[A]` - it's possible but I don't like it because it does not convey enough information, this is any function but the name `CircuitBreaker` tells me what it actually does.

Or is there something I am missing and you would solve my use case differently?

**Comment by jozic** [comment]
@jakubjanecek can you use methods in constructor to add your metrics and logging effects? 
```
  onRejected: F[Unit],
  onClosed: F[Unit],
  onHalfOpen: F[Unit],
  onOpen: F[Unit]
```

**Comment by jakubjanecek** [comment]
I can and I did but as far as I remember I had get rid of some of the features I wanted to implement. In regular application code this is just fine but if you are implementing a library with some extra features over `CircuitBreaker` then this is not enough. I understand the problem with backwards compatibility so I think we can close this.
